_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,1079 @@
+---
+title: "How to debug an ARM Cortex-M HardFault"
+description: "A step by step guide of how to investigate what led to a fault on a Cortex-M device with examples and strategies about how to recover"
+tag: [cortex-m]
+author: chris
+---
+
+At one point or another when working with an ARM Cortex-M MCU, you will hit something like:
+
+```shell
+(gdb) backtrace
+#0  HardFault_Handler() at startup.c:136
+#1  <signal handler called>
+```
+
+<!-- excerpt start -->
+
+In this article we will demystify how to determine what went wrong when an ARM Cortex-M fault occurs. We will walk step by step through what the registers to inspect and what the values actually mean. We will discuss debug tricks that can be used for some of the most challenging classes of HardFaults, how how to automate the fault analysis, and how to recover from a fault without rebooting the system. Finally, we will apply what we have learned in the article and walk through a couple practical examples.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest
+posts straight to your mailbox_
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Determining What Caused The Fault
+
+All MCUs in the Cortex-M series have several different pieces of state which can be analyzed when a fault takes place to trace down what went wrong.
+
+First we will explore the dedicated fault status registers that are present on all Cortex-M MCUs except the Cortex-M0.
+
+If you are trying to debug a Cortex-M0, you can skip ahead to [this section](#registers-prior-to-exception) where we discuss how to recover the core register state and instruction being executed at the time of the exception.
+
+> NOTE: If you already know the state to inspect when a fault occurs, you may want to skip ahead to the section about how to [automate the analysis](#automating-the-analysis).
+
+{: #cfsr}
+
+## Configurable Fault Status Registers (CFSR) - 0xE000ED28
```

Comment:
I would add a title, and fold "Configurable Fault Status Register" and "HardFault Status Register" under it. With a little blurb that could read something like:

```
## Relevant Registers

ARM Cortex-M faults can be split into two categories: configurable faults and hardfaults. By default, hardfault is the catch-all fault type and is triggered on every fault. If you would like more information, you can enable configurable faults which will invoke separate handlers for separate faults and give your code the opportunity to handle each slightly differently.

The state of hardfaults and configurable faults is exposed respectively by the HardFault Status Register, and the Configurable Fault Status Register. More info on those follows.

### Configurable Fault Status Register
[...]
### UsageFault Status Register
[...]
### HardFault Status Register
```

Reply:
Updated. fwiw, this was me playing with our levels to try and keep UsageFault, MemManage, BusFault nested under CFSR and still in the TOC ... since level 4's disappear and I think logically it makes sense for them to be nested under the CFSR